File: 117-populating-next-right-pointers-in-each-node-ii/populating_next_right_pointers_in_each_node_ii.cc
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:

    Node* connect(Node* root) {
        if(!root){
            return root;
        }

        // 建立这一层的连接
        auto p = root;
        Node* last = NULL;
        Node* head = NULL;
        
        while(p){
            
            if(p->left){
                auto & cur = p->left;
                if(last)
                    last->next = cur;
                else{
                    // 没有前面的说明当前就是头
                    head =cur;
                }
                printf("%d ",cur->val);
                last = cur;
            }
            if(p->right){
                auto & cur = p->right;
                if(last)
                    last->next = cur;
                else{
                    // 没有前面的说明当前就是头
                    head =cur;
                }
                printf("%d ",cur->val);
                last = cur;
            }
            if(p->next){
                p=p->next;
            }else{
                // 换层
                p = head;
                head = NULL;
                last = NULL;
                printf("\n");
            }
            
        }
        return root;
        
    }
};

File: 117-populating-next-right-pointers-in-each-node-ii/populating_next_right_pointers_in_each_node_ii.go
package leetcode117

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	if root == nil {
		return root
	}
	n := root
	var head *Node = nil
	var last *Node = nil
	var cur *Node
	// 巧妙利用题设，借助 next指针进行额外空间复杂度为常数的层次遍历
	// (其实是将空间转移到了每个节点上，对于普通的二叉树是无法做到的),
	// 跟着next走就相当于实现了层次遍历，借助当前的行已有的next指针进行下一行的层次遍历
	// 最开始根节点的时候直接可以进行左右节点遍历，即为层次便利
	for n != nil {
		if n.Left != nil {
			cur = n.Left
			if last == nil {
				// 第一次访问
				head = cur
			} else {
				last.Next = cur
			}
			last = cur
		}
		if n.Right != nil {
			cur = n.Right
			if last == nil {
				// 第一次访问
				head = cur
			} else {
				last.Next = cur
			}
			last = cur
		}
		n = n.Next
		if n == nil {
			// 换行
			n = head
			head = nil
			last = nil
		}

	}

	return root

}
